@babel/preset-env: Clarify `useBuiltIns` usage and purpose (#2067)
babel/website: docs/preset-env.md

`"usage"` | `"entry"` | `false`, defaults to `false`.

This option configures how `@babel/preset-env` imports various polyfills in your generated code.

Specifically, this option supports the following external polyfill packages:

> 

Reviewer: This note is useful for many users, so I don't think that remove it and add a stripped version is a good idea.

Author: I have a line below that reads:

```md
> NOTE: Previously, these would be installed via `@babel/polyfill`, which is now deprecated. The above method of installing these polyfill packages is the recommended way.
```

I don't see this as any more stripped down than the line you commented on, just moved down to be more contextual. Do you want me to move this line up?

- `core-js`: For standard JS built-ins

> - `regenerator`: For generator and `async`/`await` support

Reviewer: `regenerator` is not a polyfill.

Author: I don't think I say it is..? Unless you're saying I'm implying that it is -- in which case, could you suggest a better way of describing it?

Reviewer: > I don't think I say it is..?

> ... the following external polyfill packages:
> - `core-js`: ...
> - `regenerator`: ...
---
It's runtime, not a polyfill.

Author: Gotcha, thank you.

> - `regenerator`: For generator and `async`/`await` support

Reviewer 2: Shouldn't this be `regenerator-runtime` instead of just `regenerator`?

Author: Thank you -- fixed.

> - `regenerator-runtime`: For generator and `async`/`await` support

Reviewer 3: Note that this isn't needed for `async`/`await` if your browser supports generators.

Reviewer 3: Can you also mention this here?

Author: Isn't that assumed for this whole section, since `preset-env`'s entire purpose is to fill in for browsers' shortcomings.

i.e. `core-js` isn't needed if your browser supports all JS built-ins that it brings

Reviewer 3: My point was that `regenerator-runtime` isn't always needed by browsers that don't support `async`/`await`, while this sentence implies the opposite.

`regenerator-runtime` is only needed for browsers that don't support generators.

Author: I see what you mean, but don't know exactly what text would make sense.

Do you have a suggestion?

Reviewer 3: Maybe this?

```suggestion
- `regenerator-runtime`: For generators and `async`/`await` in browsers without generators support
```

Author: Done. Changed up your suggestion a little. I don't think everyone knows that `async`/`await` is built on generators, so I just grouped them together so it would be clear for everyone

> - `regenerator-runtime`: For generator and `async`/`await` support (for browsers that lack support)

Reviewer 3: ```suggestion
- `regenerator-runtime`: For generator and `async`/`await` support (for browsers that lack support for generators)
```

Again, it's important here to note that if a browser supports generators you can use `async` _without_ needing `regenerator-runtime` :sweat_smile:

Since these polyfills do not come with Babel or `@babel/preset-env` itself, they'll need to be installed as production-level dependencies. This way, when the Babel-generated code imports them, they'll be accessible (in `node_modules`).

> 

Reviewer: Usage of different `core-js` versions causes many questions and problems for users, so remove the note of how to install different `core-js` versions is not a good idea.

Author: That's what I was trying to do by removing the different versions. I found the fact that the documentation had two different versions of `core-js` confusing -- why are there two? Which one should I be using? It made me do unnecessary research into `core-js`, which I don't think a user of `@babel/preset-env` should have to do.

Are there specific use cases for 2 vs 3? If so, I'd be happy to add notes about them 🙂

> Since these packages do not come with Babel or `@babel/preset-env` itself, they'll need to be installed as production-level dependencies. This way, when the Babel-generated code imports them, they'll be accessible (in `node_modules`).

Reviewer 4: Wonder if we should be more explicit than "production-level", something like:

> installed a `dependencies` (not `devDependencies`)

```sh
npm install core-js regenerator-runtime --save
```

> NOTE: Previously, these would be installed via `@babel/polyfill`, which is now deprecated. The above method of installing these polyfill packages is the recommended way.

#### `useBuiltIns: 'entry'`

> NOTE: Only use `import "core-js";` and `import "regenerator-runtime/runtime";` once in your whole app.
> If you are using `@babel/polyfill` (which is deprecated), it already includes both `core-js` and `regenerator-runtime`: importing it twice will throw an error.
> Multiple imports or requires of those packages might cause global collisions and other issues that are hard to trace.
> We recommend creating a single entry file that only contains the `import` statements.

> This option enables a new plugin that replaces the `import "core-js"` and `import "regenerator-runtime/runtime"` statements with only what's needed from each of them, depending on the target environments.

Reviewer 4: ```suggestion
This option enables a plugin that replaces the `import "core-js"` and `import "regenerator-runtime/runtime"` statements with only what your target environments need from each of them.
```

**In**

```js
import "core-js";
import "regenerator-runtime/runtime";
```

**Out (different based on environment)**

```js
// If the target envs don't support these methods
import "core-js/modules/es.string.pad-start";
import "core-js/modules/es.string.pad-end";

> // If the target envs don't support generators or `async`/`await`

Reviewer 3: ```suggestion
// If the target envs don't support generators
```
